cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        MetricType.COUNTER,
+        ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, streamId.getNamespaceId(),
+                        Constants.Metrics.Tag.STREAM, streamId.getName()),
+        ImmutableList.<String>of()
+      );
+
+      try {
+        Collection<MetricTimeSeries> metrics = metricStore.query(metricDataQuery);
+        if (metrics == null || metrics.isEmpty()) {
+          // Data is not yet available, which means no data has been ingested by the stream yet
+          return new StreamSize(0L, System.currentTimeMillis());
+        }
+
+        MetricTimeSeries metric = metrics.iterator().next();
+        List<TimeValue> timeValues = metric.getTimeValues();
+        if (timeValues == null || timeValues.size() != 1) {
+          throw new IOException("Should collect exactly one time value");
+        }
+        TimeValue timeValue = timeValues.get(0);
+        // The metric store gives us 0 as the timestamp, hence we cannot use it here
+        return new StreamSize(timeValue.getValue(), System.currentTimeMillis());
+      } catch (Exception e) {
+        Throwables.propagateIfInstanceOf(e, IOException.class);
+        throw new IOException(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private StreamSizeSchedule streamSizeSchedule;
+
+    private long baseSize;
+    private long baseTs;
+    private AtomicBoolean active;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with baseSize {}, baseTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.active = new AtomicBoolean(active);
+    }
+
+    public boolean isActive() {
+      return active.get();
+    }
+
+    public void received(StreamSizeNotification notification) {
+      if (!active.get()) {
+        return;
+      }
+      long pastRunSize;
+      long pastRunTs;
+      StreamSizeSchedule currentSchedule;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen if no notification is received for a stream for some time, and we poll the stream events
+          // size using metrics, and some metric events have hit their TTL. In that case it's impossible to know
```

Comment:
While there's no doubt we want to figure out metrics data expiration (there's https://issues.cask.co/browse/CDAP-1124 started while ago that is related), it can be pretty tricky for dependent components to handle that properly. The biggest complexity with "totals", like in this case. On metrics system side, these will never be TTLed for now: https://issues.cask.co/browse/CDAP-1626.

At the same time, you may also think about handling such cases as metrics data cleanup (manual). I can see, if metrics data is pollutted and takes lots in the cluster, some may decided to manually drop it.. It is another question if we support that explicitly, but the system should not fail completely:)